src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServer/Handler/Restore/RestoreHandler.cs

```diff
@@ -0,0 +1,107 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+/// <summary>
+/// Given an input project (or none), runs restore on the project and streams the output
+/// back to the client to display.
+/// </summary>
+[ExportCSharpVisualBasicStatelessLspService(typeof(RestoreHandler)), Shared]
+[Method(MethodName)]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class RestoreHandler(DotnetCliHelper dotnetCliHelper) : ILspServiceRequestHandler<RestoreParams, RestorePartialResult[]>
+{
+    internal const string MethodName = "workspace/_roslyn_restore";
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public async Task<RestorePartialResult[]> HandleRequestAsync(RestoreParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Solution);
+        using var progress = BufferedProgress.Create(request.PartialResultToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Restore_started));
+
+        var restorePaths = GetRestorePaths(request, context.Solution, context);
+        if (restorePaths.IsEmpty)
+        {
+            progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Nothing_found_to_restore));
+            return progress.GetValues() ?? [];
+        }
+
+        await RestoreAsync(restorePaths, progress, cancellationToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, $"{LanguageServerResources.Restore_complete}{Environment.NewLine}"));
+        return progress.GetValues() ?? [];
+    }
+
+    private async Task RestoreAsync(ImmutableArray<string> pathsToRestore, BufferedProgress<RestorePartialResult> progress, CancellationToken cancellationToken)
+    {
+        foreach (var path in pathsToRestore)
+        {
+            var arguments = $"restore \"{path}\"";
+            var workingDirectory = Path.GetDirectoryName(path);
+            var stageName = string.Format(LanguageServerResources.Restoring_0, Path.GetFileName(path));
+            ReportProgress(progress, stageName, string.Format(LanguageServerResources.Running_dotnet_restore_on_0, path));
+
+            var process = dotnetCliHelper.Run(arguments, workingDirectory, shouldLocalizeOutput: true);
+
+            process.OutputDataReceived += (sender, args) => ReportProgress(progress, stageName, args.Data);
+            process.ErrorDataReceived += (sender, args) => ReportProgress(progress, stageName, args.Data);
+            process.BeginOutputReadLine();
+            process.BeginErrorReadLine();
+
+            await process.WaitForExitAsync(cancellationToken);
+
+            if (process.ExitCode != 0)
+            {
+                throw new InvalidOperationException(LanguageServerResources.Failed_to_run_restore_see_output_for_details);
```

Comment:
Should this also get reported through the other progress reporter, since I'd assume failure to restore is mostly a user problem?